---
title: "CSCB_Proj6_Code"
author: "Travis Brady"
date: "4/16/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r Library}
library(hdf5r)
library(Seurat)
library(SeuratDisk)
library(SingleCellExperiment)
library(gam)
library(scales)
library(RaceID)
library(tidyverse)
library(Matrix)
source("helper_2.R")
source("helper_6.R")
```

```{r Data}
download.file(url = 'https://raw.githubusercontent.com/czbiohub/tabula-muris-vignettes/master/data/TM_droplet_metadata.csv', 'droplet_meta.csv')
query.h5 <- LoadH5Seurat("Lu_Query_2021.h5seurat")
ref.h5 <- LoadH5Seurat("TS_Ref_2021.h5seurat")
tm.droplet.matrix <- readRDS("TM_droplet_mat.rds")
expTMraw <- utils_loadObject("~/Documents/Johns Hopkins/Courses/Computational Stem Cell Biology/Projects/CSCB_Project_6/6k_beadpurfied_raw.rda")
```

```{r More Data}
counts <- readMM(file = 'matrix.mtx')
counts <- as(counts, "dgCMatrix")
barcodes <- read.csv(file = 'barcodes.tsv', header = F)
barcodes <- t(as.vector(barcodes))
genes <- read.csv(file = 'genes.tsv', header = F, sep = '\t')
genes <- t(as.vector(genes[,1]))
rownames(counts) <- genes
colnames(counts) <- barcodes
```

```{r Create CSV w/ Raw Counts}
# write.csv(as.matrix(query.h5@assays$RNA@counts),
#           'query_counts.csv', sep  = ',',
#           row.names = T, col.names = T, quote = F)

# write.csv(as.matrix(ref.h5@assays$RNA@counts),
#           'ref_counts.csv', sep  = ',',
#           row.names = T, col.names = T, quote = F)

# write.csv(as.matrix(expTMraw),
#           'outside_counts.csv', sep  = ',',
#           row.names = T, col.names = T, quote = F)

# write.csv(as.matrix(counts),
#           'droplet_counts.csv', sep  = ',',
#           row.names = T, col.names = T, quote = F)
```

```{r Import CSV w/ Counts}

anno <- read.csv("oTab.csv")

query_counts <- read.csv("query_counts.csv", sep=",", header=TRUE)
rownames(query_counts) = query_counts[,1]

ref_counts <- read.csv("ref_counts.csv", sep=",", header=TRUE)
rownames(ref_counts) = ref_counts[,1]

out_counts <- read.csv("outside_counts.csv", sep=",", header=TRUE)
rownames(out_counts) = out_counts[,1]

drp_counts <-  read.csv("droplet_counts.csv", sep=",", header=TRUE) 
rownames(drp_counts) = drp_counts[,1]

#  prepare  data  for all pipelines by removing ERCC spike-ins, first column (names)
qryprdata <- query_counts[grep("ERCC",rownames(query_counts),invert=TRUE),-1]
refprdata  <- ref_counts[grep("ERCC",rownames(ref_counts),invert=TRUE),-1]
outprdata <- out_counts[grep("ERCC",rownames(out_counts),invert=TRUE),-1]
drpprdata <- drp_counts[grep("ERCC",rownames(drp_counts),invert=TRUE),-1]
```

# Notes

STEMID is based on [Grun et al 2016](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4985539/). It extrapolates cell fate srom single cell RNA sequencing data based on two observations/assumptions:

- stem cells produce a larger number of branches when used in TI
- the transcriptome specificity of stem cells tends to be lower (promiscuous: less defined genes)

Many types of pluripotent stem (PS) cells

- (naive) ES cells are derived from pre-implantation embryos
- (primed) EpiSCs are derived from later, epiblast tissues

# STEMID

```{r}
sc <- SCseq(qryprdata)
sc <- filterdata(sc, mintotal = 2000, minexpr = 5, minnumber = 5)
sc <- compdist(sc,metric="pearson")
sc <- clustexp(sc,cln=7,sat=FALSE)
sc <- findoutliers(sc)
sc <- comptsne(sc)
sc <- compumap(sc)
```


# Conversion for Comparison

```{r Remove low Corr Genes}
anno = anno[anno$corrs > 0.5,]
```

```{r Trim Query/Reference Counts}

# trim both data sets so that they are:
  # # using murine gene names
  # # ignore any genes that cannot be converted (i.e. low correlation or missing from 'oTab.csv')
qryTrim = convertR(query = qryprdata, dir = 'mouse')
refTrim = convertR(query = refprdata, dir = 'human')
refTrim = convertR(query = refTrim, dir = 'mouse')
outTrim = convertR(query = outprdata, dir = 'mouse')
drpTrim = convertR(query  = drpprdata, dir = 'human')
drpTrim = convertR(query  = drpTrim, dir = 'mouse')
```

```{r Report Dimensions}

# the trimmed reference data has 15302 genes, ~7500 cells
# the trimmed query data has 14291 genes, ~10k cells
dim(refTrim)
dim(qryTrim)
dim(outTrim)
dim(drpTrim)
```



# Create Fate Scores

## Find Labels

```{r Label Extraction}

# extract true labels for reference data

# rename (get rid of '-', replace with '.')
cellTypes = ref.h5$cell_type
qryTimes = query.h5$sample
ctn = names(cellTypes)
ctn = gsub("-", ".", ctn)
ctn = gsub(";", ".", ctn)
names(cellTypes) = ctn

# murine equivalents of key genes
ESkeyGenes = c('Sox2', 'Pou5f1', 'Tbx3', 'Gbx2')
```

##  Create Average Matrix

```{r AvgEXP Matrix}
# create a list with the mean expression profile for each cell type across all genes
AvgEXP = list()
allTypes = levels(cellTypes)
for(type in allTypes){
  sub = subsetR(type = type)
  avgexp = rowMeans(sub)
  AvgEXP =  append(AvgEXP, list(avgexp))
}
names(AvgEXP) = allTypes

# convert to matrix
AvgEXPmat <- matrix(unlist(AvgEXP), nrow = 15302)

# define matrix column names (cell types)
colnames(AvgEXPmat) = allTypes

# define row names (genes)
rownames(AvgEXPmat) = names(AvgEXP[[1]])

# scale/normalize data for better comparisons
AvgEXPscaled <- apply(AvgEXPmat, 2, rescale)
qryscaled <- apply(qryTrim, 2, rescale)
refscaled <- apply(refTrim, 2, rescale)
outscaled <- apply(outTrim, 2, rescale)
drpscaled <- apply(drpTrim, 2, rescale)
```

```{r Subset Tables (Brute)}
glial = subsetR(type = 'Bergmann glial cell')
CD8T = subsetR(type = 'CD8-positive, alpha-beta T cell')
ESC = subsetR(type = 'ESC')
Tcell = subsetR(type = 'T cell')
astr = subsetR(type = 'astrocyte')
peri = subsetR(type = 'brain pericyte')
endo = subsetR(type = 'endothelial cell')
epen = subsetR(type = 'ependymal cell')
int = subsetR(type = 'interneuron')
mac = subsetR(type = 'macrophage')
NKT = subsetR(type = 'mature NK T cell')
msNeur = subsetR(type = 'medium spiny neuron')
microg = subsetR(type = 'microglial cell')
epi = subsetR(type = 'neuroepithelial cell')
neuro = subsetR(type = 'neuron')
neuroSC = subsetR(type = 'neuronal stem cell')
olig = subsetR(type = 'oligodendrocyte')
oligPre = subsetR(type = 'oligodendrocyte precursor cell')
```

## Order Genes and Extract Fate

```{r Finish Up Avg, Qry, Ref and Drp Data}

# also order rows so genes appear in identical alphanumeric order

avgFin <- AvgEXPscaled[rownames(AvgEXPscaled) %in% rownames(qryscaled),]
avgFin <- avgFin[order(rownames(avgFin)),]
qryFin <- qryscaled[rownames(qryscaled) %in% rownames(AvgEXPscaled),]
qryFin <- qryFin[order(rownames(qryFin)),]
refFin <- refscaled[rownames(refscaled) %in% rownames(qryscaled),]
refFin <- refFin[order(rownames(refFin)),]
outFin <- outscaled[rownames(outscaled) %in% rownames(qryscaled),]
outFin <- outFin[order(rownames(outFin)),]
drpFin <- drpscaled[rownames(drpscaled) %in% rownames(AvgEXPscaled),]
drpFin <- drpFin[order(rownames(drpFin)),]

dim(avgFin)
dim(qryFin)
dim(refFin)
dim(outFin)
```

```{r Extract quasi-fate}

# Run only ONCE, takes  a while


#  Query

qryFate = vector(mode = 'list', length = ncol(qryFin))
for(i in 1:ncol(qryFin)){
  # print(i)
  fate = totalR(comparR(query = qryFin, cell = i))
  qryFate[i] = list(fate)
}
names(qryFate) = colnames(qryFin)

# Reference

refFate = vector(mode = 'list', length = ncol(refFin))
for(i in 1:ncol(refFin)){
  # print(i)
  fate = totalR(comparR(query = refFin, cell = i))
  refFate[i] = list(fate)
}
names(refFate) = colnames(refFin)

# Outside (1)

outFate = vector(mode = 'list', length = ncol(outFin))
outq = outFin[rownames(outFin) %in% rownames(avgFin),]
outavg = avgFin[rownames(avgFin) %in% rownames(outFin),]
for(i in 1:ncol(outFin)){
  print(i)
  fate = totalR(comparR(query = outq,avg = outavg, cell = i))
  outFate[i] = list(fate)
}
names(outFate) = colnames(outFin)

# Droplet - Outside (2)

drpFate = vector(mode = 'list', length = ncol(drpFin))
for(i in 1:ncol(drpFin)){
  # print(i)
  fate = totalR(comparR(query = drpFin, cell = i))
  drpFate[i] = list(fate)
}
names(drpFate) = colnames(drpFin)
```

```{r Save Fate Scores}
saveRDS(qryFate, file = 'Query_fate_scores.rds')
saveRDS(refFate, file = 'Reference_fate_scores.rds')
saveRDS(outFate, file = 'Outside_fate_scores.rds')
saveRDS(drpFate, file = 'Droplet_fate_scores.rds')
```

# Results

```{r Load Fate Scores}
qryFate = readRDS(file = 'Query_fate_scores.rds')
refFate = readRDS(file = 'Reference_fate_scores.rds')
outFate = readRDS(file = 'Outside_fate_scores.rds')
drpFate = readRDS(file = 'Droplet_fate_scores.rds')
```

```{r Extract Labels for Each Time in Query}

# each of these (D0, D1, etc. in query) is the list of names with that specific label

# some rownames were made to drop underscores and replace with dashes, so introduced a mismatch in names, should have  done this earlier with the full list of names, but  fixing it here

levels(qryTimes)
D0 = names(qryTimes[qryTimes == "IMR90_iBMEC_D0"])
D0 = gsub("-", ".", D0)

D3 = names(qryTimes[qryTimes == "IMR90_iBMEC_D3"])
D3 = gsub("-", ".", D3)

D6 = names(qryTimes[qryTimes == "IMR90_iBMEC_D6"])
D6 = gsub("-", ".", D6)

D11 = names(qryTimes[qryTimes == "IMR90_iBMEC_D11"])
D11 = gsub("-", ".", D11)

D16 = names(qryTimes[qryTimes == "IMR90_iBMEC_D16"])
D16 = gsub("-", ".", D16)

D50 = names(qryTimes[qryTimes == "IMR90_iBMEC_D50"])
D50 = gsub("-", ".", D50)

rEC = names(qryTimes[qryTimes == "IMR90_rEC"])
rEC = gsub("-", ".", rEC)

BMEC = names(qryTimes[qryTimes == "primary_BMEC"])
BMEC = gsub("-", ".", BMEC)

HUVEC = names(qryTimes[qryTimes == "HUVEC"])
HUVEC = gsub("-", ".", HUVEC)
```

```{r Extract Labels for Each Type in Reference}

# each of these (glial, CD8T, etc. in reference) is the list of names with to that specific label

glial = names(cellTypes[cellTypes == 'Bergmann glial cell'])
CD8T = names(cellTypes[cellTypes == 'CD8-positive, alpha-beta T cell'])
ESC = names(cellTypes[cellTypes == 'ESC'])
Tcell = names(cellTypes[cellTypes == 'T cell'])
astr = names(cellTypes[cellTypes == 'astrocyte'])
peri = names(cellTypes[cellTypes == 'brain pericyte'])
endo = names(cellTypes[cellTypes == 'endothelial cell'])
epen = names(cellTypes[cellTypes == 'ependymal cell'])
interneur = names(cellTypes[cellTypes == 'interneuron'])
macro = names(cellTypes[cellTypes == 'macrophage'])
NKT = names(cellTypes[cellTypes == 'mature NK T cell'])
spinyneur = names(cellTypes[cellTypes == 'medium spiny neuron'])
microgl = names(cellTypes[cellTypes == 'microglial cell'])
neuro = names(cellTypes[cellTypes == 'neuron'])
neuroSC = names(cellTypes[cellTypes == 'neuronal stem cell'])
oligo = names(cellTypes[cellTypes == 'oligodendrocyte'])
preoligo = names(cellTypes[cellTypes == 'oligodendrocyte precursor cell'])
epineur = names(cellTypes[cellTypes == 'neuroepithelial cell'])
```

```{r Separate by Day in query and Type in reference}

# find the names of cells that fit each category
  # day/type in query
  # type in reference

# Query
D0_exp = qryFate[names(qryFate) %in% D0]
HUVEC_exp = qryFate[names(qryFate) %in% HUVEC]
D3_exp = qryFate[names(qryFate) %in% D3]
D6_exp = qryFate[names(qryFate) %in% D6]
D11_exp = qryFate[names(qryFate) %in% D11]
D16_exp = qryFate[names(qryFate) %in% D16]
D50_exp = qryFate[names(qryFate) %in% D50]
rEC_exp = qryFate[names(qryFate) %in% rEC]
BMEC_exp = qryFate[names(qryFate) %in% BMEC]

# Reference
glial_exp = refFate[names(refFate) %in% glial]
CD8T_exp = refFate[names(refFate) %in% CD8T]
ESC_exp = refFate[names(refFate) %in% ESC]
Tcell_exp = refFate[names(refFate) %in% Tcell]
astr_exp = refFate[names(refFate) %in% astr]
peri_exp = refFate[names(refFate) %in% peri]
endo_exp = refFate[names(refFate) %in% endo]
epen_exp = refFate[names(refFate) %in% epen]
interneur_exp = refFate[names(refFate) %in% interneur]
macro_exp = refFate[names(refFate) %in% macro]
NKT_exp = refFate[names(refFate) %in% NKT]
spinyneur_exp = refFate[names(refFate) %in% spinyneur]
microgl_exp = refFate[names(refFate) %in% microgl]
epineur_exp = refFate[names(refFate) %in% epineur]
neuro_exp = refFate[names(refFate) %in% neuro]
neuroSC_exp = refFate[names(refFate) %in% neuroSC]
oligo_exp = refFate[names(refFate) %in% oligo]
preoligo_exp = refFate[names(refFate) %in% preoligo]
```

```{r Average Scores}
QRT = c(averageR(scor = HUVEC_exp, typ = 'ESC'),
        averageR(scor = D0_exp, typ = 'ESC'),
        averageR(scor = D3_exp, typ = 'ESC'),
        averageR(scor = D6_exp, typ = 'ESC'),
        averageR(scor = D11_exp, typ = 'ESC'),
        averageR(scor = D16_exp, typ = 'ESC'),
        averageR(scor = D50_exp, typ = 'ESC'),
        averageR(scor = rEC_exp, typ = 'ESC'),
        averageR(scor = BMEC_exp, typ = 'ESC'))
names(QRT) = levels(qryTimes)

REFT = c(averageR(scor = glial_exp, typ = 'ESC'),
         averageR(scor = CD8T_exp, typ = 'ESC'),
         averageR(scor = ESC_exp, typ = 'ESC'),
         averageR(scor = Tcell_exp, typ = 'ESC'),
         averageR(scor = astr_exp, typ = 'ESC'),
         averageR(scor = peri_exp, typ = 'ESC'),
         averageR(scor = endo_exp, typ = 'ESC'),
         averageR(scor = epen_exp, typ = 'ESC'),
         averageR(scor = interneur_exp, typ = 'ESC'),
         averageR(scor = macro_exp, typ = 'ESC'),
         averageR(scor = NKT_exp, typ = 'ESC'),
         averageR(scor = spinyneur_exp, typ = 'ESC'),
         averageR(scor = microgl_exp, typ = 'ESC'),
         averageR(scor = epineur_exp, typ = 'ESC'),
         averageR(scor = neuro_exp, typ = 'ESC'),
         averageR(scor = neuroSC_exp, typ = 'ESC'),
         averageR(scor = oligo_exp, typ = 'ESC'),
         averageR(scor = preoligo_exp, typ = 'ESC'))
names(REFT) = levels(cellTypes)
```

```{r Plot Query Data as Points}
qry_bind = cbind(QRT[2:7], c(0,3,6,11,16,50))
colnames(qry_bind) = c('score', 'day')
qry_bind = as.data.frame(qry_bind)

query_plots <- ggplot(data = qry_bind, mapping = aes(x = day, y = score))
query_plots + geom_point(color = 'blue')

dev.copy(pdf, 'query_scores.pdf')
dev.off()
```

```{r Plot Reference Data as Bars}
ref_bind = cbind(REFT, names(REFT))
colnames(ref_bind) = c('score', 'type')
ref_bind = as.data.frame(ref_bind)
ref_bind$score = as.numeric(ref_bind$score)
ref_bind$type <- factor(ref_bind$type, levels = ref_bind[order(ref_bind$score, decreasing = T),]$type)
# ref_bind$type <- factor(ref_bind$type, levels = ref_bind[order(ref_bind$score),])
# positions = rownames(ref_bind[order(ref_bind$score, decreasing = T),])

ref_plots <- ggplot(data = ref_bind, aes(x = type, y = score))
ref_plots +
  geom_bar(stat = 'identity', fill = 'blue') +
  theme(axis.text.x = element_text(angle = 80, vjust = 0.95, hjust=1)) +
  scale_y_continuous(breaks=seq(0, 0.3, 0.05))

dev.copy(pdf, 'ref_scores.pdf')
dev.off()
```



